src/basler_synced_cameras_node.cpp

```diff
@@ -0,0 +1,293 @@
+#include <ros/console.h>
+#include <ros/ros.h>
+
+#include <pylon/PylonIncludes.h>
+#include <pylon/gige/BaslerGigEInstantCamera.h>
+
+#include <XmlRpcValue.h>
+
+#include "basler_parameters.h"
+#include "image_publisher.h"
+
+std::vector<Pylon::CBaslerGigEInstantCamera> loadCameras(const XmlRpc::XmlRpcValue& cameras_yaml)
+{
+    Pylon::CTlFactory& tl_factory = Pylon::CTlFactory::GetInstance();
+    Pylon::DeviceInfoList_t devices;
+    ros::Rate r(0.5);
+
+    std::vector<Pylon::CBaslerGigEInstantCamera> cameras(cameras_yaml.size());
+    for (size_t index = 0; index < cameras_yaml.size(); ++index)
+    {
+        ROS_ASSERT_MSG(cameras_yaml[index].getType() == XmlRpc::XmlRpcValue::TypeStruct, "Bad yaml for camera %zu", index);
+        ROS_ASSERT_MSG(cameras_yaml[index].hasMember("serial"), "Camera %zu yaml missing serial", index);
+        std::string serial = std::to_string(int(cameras_yaml[index]["serial"]));
+
+        bool camera_found = false;
+        ros::Time timeout_time = ros::Time::now() + ros::Duration(5);
+        while(ros::ok() && !camera_found)
+        {
+            tl_factory.EnumerateDevices(devices);
+            for (size_t i = 0; i < devices.size(); ++i)
+            {
+                if (devices[i].GetSerialNumber().c_str() == serial)
+                {
+                    ROS_INFO_STREAM("Found camera " << index << " with matching serial " << serial);
+                    cameras[index].Attach(tl_factory.CreateDevice(devices[i]));
+                    camera_found = true;
+                    break;
+                }
+            }
+            if (!camera_found)
+            {
+                ROS_DEBUG_STREAM("Failed to find camera " << index << " with serial " << serial);
+                ROS_DEBUG_STREAM( devices.size() << " other serials found:");
+                for (size_t i = 0; i < devices.size(); ++i)
+                {
+                    ROS_DEBUG_STREAM(devices[i].GetSerialNumber());
+                }
+                if (ros::Time::now() > timeout_time)
+                {
+                    ROS_ERROR_STREAM("Timed out looking for camera " << index << " with serial " << serial);
+                    return {};
+                }
+                r.sleep();
+                ros::spinOnce();
+            }
+        }
+    }
+    return cameras;
+}
+
+// Wait for all cameras to be slaves and to have same master clock id
+bool waitForPTPSlave(const std::vector<Pylon::CBaslerGigEInstantCamera>& cameras)
+{
+    ROS_DEBUG("Waiting for %zu cameras to be PTP slaves...", cameras.size());
+    ros::Rate r(0.5);
+    ros::Time timeout_time = ros::Time::now() + ros::Duration(60);
```

Comment:
Could you set these default timeouts via a constant variable at the top of the file (or from a config file if there is one)?  I'm not sure they need to be passed in through the launch file (though that might be helpful if you think users might want to change the timeouts).